### ISSUE TITLE
Port to WebKit2 4.0 API

### ISSUE THREAD

**Issue by quozl** [sugarlabs/read-activity#17]
Read uses for EPUB viewing the WebKit 3.0 API, based on WebKit 2.4, which is unmaintained and dangerous.  Port to WebKit2 4.0 API based on WebKit 2.16 or later.

**Comment by quozl** [comment]
Quick check of downstreams;
* Debian, does [declare](https://anonscm.debian.org/cgit/pkg-sugar/sugar-read-activity.git/tree/debian/rules#n43) the WebKit dependency, and they have bug [866649](https://bugs.debian.org/866649),
* Fedora, the [spec file](https://src.fedoraproject.org/cgit/rpms/sugar-read.git/tree/sugar-read.spec) does not declare a WebKit dependency, no bug found, and on Fedora 25 removing the webkitgtk3 package does cause Read to fail.

**Comment by jbicha** [comment]
Fedora 27 (expected to release in about November) is also removing the old webkitgtk.

**Comment by Keepco** [comment]
If you already work on the port would you mind creating a PR or a project for it? Just to avoid work being done twice :)

**Comment by quozl** [comment]
@Keepco, good idea.  I've done nothing significant on it.  Two mailing list posts;

* [Call for porting](http://lists.sugarlabs.org/archive/sugar-devel/2017-July/054455.html) and my intention to remove the EPUB feature on 2nd August if there is no response,
* [Walter says Ayush wants to work on it](http://lists.sugarlabs.org/archive/sugar-devel/2017-July/054472.html) but nothing public since.

Time is short, and I'm busy.  I'll review patches from anyone.

**Comment by quozl** [comment]
Discussion with Ayush in #sugar IRC:

`<ayushhere>` Hi Quozl

`<Quozl>` g'day.

`<ayushhere>` Quozl: I wanted to discuss some stuff about the task, as in, I know a few things about webkit but not really sure how to work with the WebKit2 4.0

`<ayushhere>` tried finding a few tutorials but couldn't find a lot about it

`<Quozl>` WebKit is a browser.  WebKit2 is an API.  Know what an API is?

`<ayushhere>` API is functions and tools used together

`<Quozl>` More or less, yes.  Language environment of Read is Python with GObject and GTK+ API calls, so the best binding to use is the Python GI API or PyGObject API for WebKit2.

`<Quozl>` At the moment, Read is using WebKit 3.0 API, which you should be able to confirm by reading the source code of Read.

`<ayushhere>` Yeah, I have gone through the code

`<Quozl>` Each PyGObject API is self-documenting, and the documentation can be generated on any system using variouos tools, or you can use someone else's generated HTML, such as https://lazka.github.io/pgi-docs/

`<Quozl>` You'll see in that URL there is documentation for both WebKit 3.0 API and WebKit2 4.0 API.

`<Quozl>` ayushhere: there are several examples of code in Sugar and activities that use WebKit2 4.0 API, such as View Help and Browse.

`<ayushhere>` Quozl: ok, will see those

`<Quozl>` ayushhere: is that all you wanted to know?  two ways to work with an API; write code after reading documentation, and write code after reading other code that uses same API.

`<ayushhere>` Quozl: I guess so, I'm doing through the documentation right now and I'll see the codes as well. Will let you know how it goes

**Comment by quozl** [comment]
Nobody seems to be working on this, so I'm looking into it now; either removing EPUB support or porting.

**Comment by quozl** [comment]
https://github.com/sugarlabs/read-activity/pull/18 removes EPUB support.  Not the best fix for the problem.

**Comment by quozl** [comment]
After a few days trying to port, I'm giving up.

Last remaining issue is that the EPUB viewer renders each file in the EPUB as a different web page, and was reliant on WebKit accepting a Gtk.Scrollbar wrapper widget, which WebKit2 does not accept, so while I've got content visible it can't be easily scrolled except with mouse scroll buttons, and the page number doesn't update.

So I've merged https://github.com/sugarlabs/read-activity/pull/18

If anyone is available to work this problem further, I can push my work in progress branch.
